scipy/fft/tests/test_helper.py

```diff
@@ -391,22 +391,13 @@ def test_uneven_dims(self, xp):
         xp_assert_close(fft.ifftshift(shift_dim_both), freqs)
 
 
-@skip_if_array_api('array_api_strict', 'cupy',
-                   reasons=['fft not yet implemented by array-api-strict',
-                            'cupy.fft not yet implemented by array-api-compat'])
+@skip_if_array_api('cupy',
+                   reasons=['cupy.fft not yet implemented by array-api-compat'])
 class TestFFTFreq:
 
     def test_definition(self, xp):
-        device = SCIPY_DEVICE
```

Comment:
I'm not sure what setting the device was achieving here? If anything, what would need to be tested is the `device` parameter of `fftfreq`, right? But I'm not sure if that is possible, without a common device shared by all backends (or a big `if` statement / dictionary). Maybe we do want to keep a dictionary of default devices for each backend we test with for this purpose?

Reply:
Looks to me like that was needed to make the tests work. In `scipy/conftest.py` we have:
```python
torch.set_default_device(SCIPY_DEVICE)
```
so with a creation function like `fftfreq`, the result may end up on GPU. Hence `x`/`x2` to compare against the result of calling `fftfreq` may need to be on GPU as well.

Did you try this change with pytorch-cuda?

Reply:
Does the array API standard mandate the ability to set the default device globally? Not sure if that matters for design here.

I did some local testing on NVIDIA GPU. One problem is that unlike you and Ralf, I consistently see threading-related errors with FFT in the tessuite (we discussed this before), so I have to establish a baseline every time I run the tests for array API. I seem to have one extra fft failure on this branch vs. `main` below, but I think it is just another threading-related issue rather than something specific to this?

I'm also assuming you don't need me to test against NumPy `main` because of the external strict lib now?

On SciPy `main`, with NumPy `1.26.4`, ` SCIPY_DEVICE=cuda python dev.py test -j 32 -b all`:

<details>

```
FAILED scipy/fft/tests/test_basic.py::TestFFT1D::test_dtypes_real[float32-torch] - AssertionError: Tensor-likes are not close!
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_fft[torch] - pytest.PytestUnhandledThreadExceptionWarning: Exception in thread Thread-33 (worker)
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_rfft[cupy] - AssertionError: 
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_ihfft[cupy] - AssertionError: 
======================================================= 4 failed, 48743 passed, 10671 skipped, 149 xfailed, 12 xpassed in 60.62s (0:01:00) ========================================================

```

</details>

On this branch, with NumPy `1.26.4`, same incantation:

<details>

```
===================================================================================== short test summary info =====================================================================================
FAILED scipy/fft/tests/test_basic.py::TestFFT1D::test_dtypes_real[float32-torch] - AssertionError: Tensor-likes are not close!
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_fft[torch] - pytest.PytestUnhandledThreadExceptionWarning: Exception in thread Thread-33 (worker)
FAILED scipy/sparse/linalg/_eigen/tests/test_svds.py::Test_SVDS_LOBPCG::test_svds_parameter_tol - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_ifft[cupy] - AssertionError: 
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_rfft[cupy] - AssertionError: 
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_ihfft[cupy] - AssertionError: 
============================================================ 6 failed, 48741 passed, 10669 skipped, 149 xfailed, 12 xpassed in 56.67s =============================================================

```

</details>

Second trial of that attempt removes the intermittent failure (gh-20157), but seems to preserve the 1 extra fft fail vs. `main`:

<details>

```
FAILED scipy/fft/tests/test_basic.py::TestFFT1D::test_dtypes_real[float32-torch] - AssertionError: Tensor-likes are not close!
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_fft[torch] - pytest.PytestUnhandledThreadExceptionWarning: Exception in thread Thread-33 (worker)
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_ifft[cupy] - AssertionError: 
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_rfft[cupy] - AssertionError: 
FAILED scipy/fft/tests/test_basic.py::TestFFTThreadSafe::test_ihfft[cupy] - AssertionError: 
```

</details>

The extra failure doesn't look related (?) though, more thread-safety stuff:

<details>

```
________________________________________________________________________________ TestFFTThreadSafe.test_ifft[cupy] ________________________________________________________________________________
[gw10] linux -- Python 3.11.2 /home/treddy/python_venvs/py_311_scipy_dev/bin/python
scipy/fft/tests/test_basic.py:440: in test_ifft
    self._test_mtsame(fft.ifft, a, xp=xp)
        a          = array([[1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+...  [1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+0.j]])
        self       = <scipy.fft.tests.test_basic.TestFFTThreadSafe object at 0x7f0207de2bd0>
        xp         = <module 'cupy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'>
scipy/fft/tests/test_basic.py:429: in _test_mtsame
    xp_assert_equal(
        args       = (array([[1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1....[1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+0.j],
       [1.+0.j, 1.+0.j, 1.+0.j, ..., 1.+0.j, 1.+0.j, 1.+0.j]]),)
        expected   = array([[1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+...  [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j]])
        func       = <uarray multimethod 'ifft'>
        i          = 1
        q          = <queue.Queue object at 0x7f01f3d17e10>
        self       = <scipy.fft.tests.test_basic.TestFFTThreadSafe object at 0x7f0207de2bd0>
        t          = [<Thread(Thread-113 (worker), stopped 139644612109888)>, <Thread(Thread-114 (worker), stopped 139644578539072)>, <Thre...>, <Thread(Thread-117 (worker), stopped 139646109480512)>, <Thread(Thread-118 (worker), stopped 139644737943104)>, ...]
        worker     = <function TestFFTThreadSafe._test_mtsame.<locals>.worker at 0x7f01f2cd4720>
        xp         = <module 'cupy' from '/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/__init__.py'>
/home/treddy/python_venvs/py_311_scipy_dev/lib/python3.11/site-packages/cupy/testing/_array.py:91: in assert_array_equal
    numpy.testing.assert_array_equal(
        err_msg    = 'Function returned wrong value in multithreaded context'
        strides_check = False
        verbose    = True
        x          = array([[0.005+0.j, 0.005+0.j, 0.005+0.j, ..., 0.005+0.j, 0.005+0.j,
        0.005+0.j],
       [0.005+0.j, 0.005+0.j, ...0.005+0.j,
        0.005+0.j],
       [0.005+0.j, 0.005+0.j, 0.005+0.j, ..., 0.005+0.j, 0.005+0.j,
        0.005+0.j]])
        y          = array([[1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+...  [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j]])
../../../../../spack/opt/spack/linux-ubuntu22.04-skylake/gcc-11.3.0/python-3.11.2-4qncg2nqev5evxfmamde3e6rnb34b4ls/lib/python3.11/contextlib.py:81: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not equal
E   Function returned wrong value in multithreaded context
E   Mismatched elements: 160000 / 160000 (100%)
E   Max absolute difference: 0.995
E   Max relative difference: 0.995
E    x: array([[0.005+0.j, 0.005+0.j, 0.005+0.j, ..., 0.005+0.j, 0.005+0.j,
E           0.005+0.j],
E          [0.005+0.j, 0.005+0.j, 0.005+0.j, ..., 0.005+0.j, 0.005+0.j,...
E    y: array([[1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
E          [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
E          [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],...
        args       = (<built-in function eq>, array([[0.005+0.j, 0.005+0.j, 0.005+0.j, ..., 0.005+0.j, 0.005+0.j,
        0.005+0.j],
     ... [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [1.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j]]))
        func       = <function assert_array_compare at 0x7f02cb2785e0>
        kwds       = {'err_msg': 'Function returned wrong value in multithreaded context', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x7f02cb264810>

```

</details>

Reply:
> Does the array API standard mandate the ability to set the default device globally?

It doesn't.

> I consistently see threading-related errors with FFT in the tessuite (we discussed this before)

I have a terrible memory, but it sounds like we should get that fixed at some point.

Reply:
I checked that I could reproduce the 5 baseline failures on `main`, so not related to this PR.

Our old discussion starts here, but maybe that's best dealt with elsewhere: https://github.com/scipy/scipy/pull/19005#issuecomment-1664647943

Reply:
I suppose we'll have to skip the ThreadSafe tests for now. Is the other one just a tolerance bump, or...

Reply:
The `SCIPY_DEVICE` was needed with PyTorch 2.0 because `torch.asarray` didn't honor `torch.set_default_device`. That was fixed in Aug'23, so this change is fine now - resolving.